async-iterator.md

```diff
@@ -0,0 +1,904 @@
+| `IAsyncEnumerator<T>` used with `await MoveNextAsync()` and `Current` | `IEnumerator<T>` used with `MoveNext()` and `Current` |
+| `IAsyncDisposable` used with `await DisposeAsync()` | `IDisposable` used with `Dispose()` |
+
+So it is unsurprising that the lowering for `await foreach` is very similar to that of a `foreach`.
+
+The following:
+```C#
+await foreach (var value in asyncEnumerable)
+{
+   ... body using `value` ...
+}
+```
+translates to:
+```C#
+var enumerator = asyncEnumerable.GetAsyncEnumerator();
+try
+{
+    while (await enumerator.MoveNextAsync())
+    {
+        var value = enumerator.Current;
+        ... body using `value` ...
+    }
+}
+finally
+{
+    await enumerator.DisposeAsync();
+}
+```
+
+
+### Control flow (caller and background execution)
+
+In a `await foreach`, the caller repeatedly calls `MoveNextAsync()`.
+This returns a `ValueTask<bool>` where the `bool` represents the presence or absence of an item.
+When you call `MoveNextAsync()`:
+- if the code reaches a `yield return` statement, you immediately get a `true`,
+- if the code reaches the end of the method, you immediately get a `false`,
+- if you reach an `await`, you get a pending task and the code continues to execute in the background.
+
+As the code continues to execute in the background:
+- if it reaches a `yield return`, the background execution completes and the task we've previously handed to the caller is fulfilled with `true`,
+- if it reaches the end of the method, that task is fulfilled with `false`,
+- if it reaches an `await`, execution continues in the background.
+
+Note that the method is moved forward either by the caller or by background execution, but never both. Calling `MoveNextAsync()` before the task from the previous call complets yield unspecified results.
+
+
+### Exception handling (optional)
+
+When the caller is moving the method forward and an exception is thrown (without being caught by user code in the async-iterator method), `MoveNextAsync()` will return a task as normal.
+
+That task is complete (rather than pending). When the caller tries to access its result the task will throw the exception it holds.
+
+When background execution is moving the method forward and an exception is thrown without handling, the exception is similarly caught and passed onto the caller via the task.
+
+
+### Parallels with iterator and async control flow (optional)
+
+The execution of iterator methods shifts the control between the caller and the method:
+- the caller repeatedly calls `MoveNext()` (which executes the method),
+- the method yields control back to the caller (when reaching a `yield return`).
+
+The execution of async methods starts by the caller giving control to the method and background execution.
+From there on, the control shifts between the method and background execution:
+- the background execution resumes the method after a long-running task completes,
```

Comment:
"long-running task"... "LongRunning" is a specific concept in the Task APIs, so I'd avoid using that term here.  Background execution resumes the method after an awaited task completes asynchronously.